client/src/main/java/io/pravega/client/stream/impl/ReaderGroupStateManager.java

```diff
@@ -462,7 +472,27 @@ void checkpoint(String checkpointName, PositionInternal lastPosition) throws Rea
     boolean isCheckpointSilent(String atCheckpoint) {
         return sync.getState().getCheckpointState().isCheckpointSilent(atCheckpoint);
     }
-    
+
+    @SneakyThrows(CheckpointFailedException.class)
+    public void updateTruncationStreamCutIfNeeded() {
+        ReaderGroupState state = sync.getState();
+        // here add check if RG is Subscriber and autoPublishTruncationStreamCut = true
+        if (state.getConfig().getRetentionConfig().equals(ReaderGroupConfig.ReaderGroupRetentionConfig.TRUNCATE_AT_LAST_CHECKPOINT)
+        && !state.getCheckpointState().isLastCheckpointPublished()) {
+            // we get here only when this is the reader that has completed the lastCheckpoint
+            Optional<Map<Stream, Map<Segment, Long>>> cuts = state.getPositionsForLastCompletedCheckpoint();
+            if (!cuts.isPresent()) {
+                throw new CheckpointFailedException("Could not get positions for last checkpoint.");
+            }
+            Set<StreamCut> streamCuts = cuts.get().entrySet().stream()
+                    .map(entry -> new StreamCutImpl(entry.getKey(), entry.getValue()))
```

Comment:
map followed by foreach is two passes.. 
we can simply do it in one pass with foreach where you do the transformation and then call update subscriber streamcut.

Reply:
Ok. fixed